core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java

```diff
@@ -0,0 +1,212 @@
+import javax.crypto.spec.SecretKeySpec;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public class AesGcmInputStream extends SeekableInputStream {
+  private SeekableInputStream sourceStream;
+  private long netSourceFileSize;
+
+  private Cipher gcmCipher;
+  private SecretKey key;
+  private byte[] nonce;
+
+  private byte[] ciphertextBlockBuffer;
+  private int cipherBlockSize;
+  private int plainBlockSize;
+  private long plainStreamPosition;
+  private int currentBlockIndex;
+  private int currentOffsetInPlainBlock;
+  private int numberOfBlocks;
+  private int lastBlockSize;
+  private long plainStreamSize;
+  private byte[] fileAadPrefix;
+
+  AesGcmInputStream(SeekableInputStream sourceStream, long sourceLength,
+                    byte[] aesKey, byte[] fileAadPrefix) throws IOException {
+    this.netSourceFileSize = sourceLength - AesGcmOutputStream.PREFIX_LENGTH;
+    this.sourceStream = sourceStream;
+    byte[] prefixBytes = new byte[AesGcmOutputStream.PREFIX_LENGTH];
+    int fetched = sourceStream.read(prefixBytes);
+    Preconditions.checkArgument(fetched == AesGcmOutputStream.PREFIX_LENGTH,
+        "Insufficient read " + fetched);
+    this.plainStreamPosition = 0;
+    this.fileAadPrefix = fileAadPrefix;
+
+    byte[] magic = new byte[AesGcmOutputStream.MAGIC_ARRAY.length];
+    System.arraycopy(prefixBytes, 0, magic, 0, AesGcmOutputStream.MAGIC_ARRAY.length);
+
+    Preconditions.checkArgument(Arrays.equals(AesGcmOutputStream.MAGIC_ARRAY, magic),
+        "File with wrong magic string. Should start with " + AesGcmOutputStream.MAGIC_STRING);
+
+    plainBlockSize = ByteBuffer.wrap(prefixBytes, AesGcmOutputStream.MAGIC_ARRAY.length, 4)
+        .order(ByteOrder.LITTLE_ENDIAN).getInt();
+    cipherBlockSize = plainBlockSize + AesGcmOutputStream.GCM_NONCE_LENGTH + AesGcmOutputStream.GCM_TAG_LENGTH;
+
+    try {
+      gcmCipher = Cipher.getInstance("AES/GCM/NoPadding");
+    } catch (GeneralSecurityException e) {
+      throw new IOException(e);
+    }
+    this.nonce = new byte[AesGcmOutputStream.GCM_NONCE_LENGTH];
+    this.key = new SecretKeySpec(aesKey, "AES");
+    this.ciphertextBlockBuffer = new byte[cipherBlockSize];
+    this.currentBlockIndex = 0;
+    this.currentOffsetInPlainBlock = 0;
+
+    numberOfBlocks = (int) (netSourceFileSize / cipherBlockSize);
+    lastBlockSize = (int) (netSourceFileSize % cipherBlockSize);
+    if (lastBlockSize == 0) {
+      lastBlockSize = cipherBlockSize;
+    } else {
+      numberOfBlocks += 1;
+    }
+
+    plainStreamSize = (numberOfBlocks - 1L) * plainBlockSize +
+            (lastBlockSize - AesGcmOutputStream.GCM_NONCE_LENGTH - AesGcmOutputStream.GCM_TAG_LENGTH);
+  }
+
+  public long plaintextStreamSize() {
+    return plainStreamSize;
+  }
+
+  @Override
+  public int available() throws IOException {
+    return Math.toIntExact(plainStreamSize - plainStreamPosition);
```

Comment:
oh, no, streams typically use long for these, eg
```java
@Override
  public void seek(long newPos)
```

Reply:
interesting though, the `available` doesn't have to be accurate, according to the InputStream contract. I presume this means that we don't need to use Math.toIntExact since it throws an exception if the result can't be mapped to an int. I'll change the logic to return max int in that case.